### PR TITLE
Updating SpinButton to a function component within React-Next

### PR DESCRIPTION
**Pull request by czearing**
#### Pull request checklist
- [X] Include a change request file using `$ yarn change`

#### Description of changes

1. Updating `SpinButton` to a function component
    - Updated to correctly accept user-entered values when value prop is set
    - Removed deprecated test
    - Updated styling
2. Added `useSetTimeOut` hook 

#### Focus areas to test
`SpinButton`

### CODE REVIEW

File: packages/react-next/src/components/SpinButton/SpinButton.base.tsx [hunk 1]
```diff
@@ -1,5 +1,5 @@
 import * as React from 'react';
-import { IconButton } from '../../Button';
+import { IconButton } from '@fluentui/react-next/lib/compat/Button';
```
Comment (khmakoto) [inline packages/react-next/src/components/SpinButton/SpinButton.base.tsx:2]:
```suggestion
import { IconButton } from '../../compat/Button';
```

File: packages/react-next/src/components/SpinButton/SpinButton.styles.ts [hunk 2]
```diff
@@ -6,7 +6,7 @@ import {
   IconFontSizes,
   getInputFocusStyle,
 } from '../../Styling';
-import { IButtonStyles } from '../../Button';
+import { IButtonStyles } from '@fluentui/react-next/lib/compat/Button';
```
Comment (khmakoto) [inline packages/react-next/src/components/SpinButton/SpinButton.styles.ts:9]:
```suggestion
import { IButtonStyles } from '../../compat/Button';

```

File: change/@fluentui-react-next-2020-04-30-15-57-11-feat-spin-button.json [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{
+  "type": "prerelease",
+  "comment": "Updating SpinButton to a function component within React-Next.",
```
Comment (khmakoto) [inline change/@fluentui-react-next-2020-04-30-15-57-11-feat-spin-button.json:3]: nit:
```suggestion
  "comment": "SpinButton: Updating to a function component within react-next.",
```

File: packages/react-next/src/components/SpinButton/SpinButton.test.tsx [hunk 4]
```diff
@@ -162,7 +163,7 @@ describe('SpinButton', () => {
     const inputDOM = wrapper.getDOMNode().querySelector('input')!;
     expect(inputDOM.value).toBe('');
     expect(inputDOM.getAttribute('aria-valuenow')).toBe(null);
-    expect(inputDOM.getAttribute('aria-valuetext')).toBe('');
+    expect(inputDOM.getAttribute('aria-valuetext')).toBe(null);
```
Comment (khmakoto) [inline packages/react-next/src/components/SpinButton/SpinButton.test.tsx:166]: I think this is actually correct but I wonder at what made you realize this needed to be changed?

File: packages/react-next/src/components/SpinButton/SpinButton.test.tsx [hunk 5]
```diff
@@ -282,18 +283,32 @@ describe('SpinButton', () => {
...
     expect(inputDOM.value).toBe('21');
     expect(inputDOM.getAttribute('aria-valuenow')).toBe('21');
   });
 
+  it('does not accept user-entered values when controlled', () => {
+    wrapper = mount(<SpinButton componentRef={ref} value="12" />);
+
+    const inputDOM = wrapper.getDOMNode().querySelector('input')!;
+    ReactTestUtils.Simulate.input(inputDOM, mockEvent('21'));
+    ReactTestUtils.Simulate.blur(inputDOM);
```
Comment (khmakoto) [inline packages/react-next/src/components/SpinButton/SpinButton.test.tsx:305]: Should you be putting this 2 inside an `act` call as well?

File: change/@uifabric-file-type-icons-2020-06-22-11-41-59-caperez-filetype_cortex_and_shortcut.json [hunk 6]
```diff
@@ -0,0 +1,8 @@
+{
+  "type": "patch",
+  "comment": "revving CDN url and adding .classifier type to the filetype list for icons",
```
Comment (xugao) [inline change/@uifabric-file-type-icons-2020-06-22-11-41-59-caperez-filetype_cortex_and_shortcut.json:3]: you have some change files here that shouldn't be added

File: packages/react-next/src/components/SpinButton/SpinButton.base.tsx [hunk 7]
```diff
@@ -0,0 +1,452 @@
...
+
+  const onIncrementMouseDown = (): void => {
+    updateValue(true /* shouldSpin */, state.initialStepDelay, onIncrement!);
+  };
+
+  const onDecrementMouseDown = (): void => {
+    updateValue(true /* shouldSpin */, state.initialStepDelay, onDecrement!);
+  };
+
+  useComponentRef(props, input, value);
```
Comment (xugao) [inline packages/react-next/src/components/SpinButton/SpinButton.base.tsx:349]: nit: let move this up where other hook calls are

File: packages/react-next/src/components/SpinButton/SpinButton.base.tsx [hunk 8]
```diff
@@ -0,0 +1,452 @@
...
+
+  const [value, setValue] = useControllableValue(
+    props.value,
+    props.defaultValue !== undefined ? props.defaultValue : String(min || '0'),
+  );
+
+  // keyboard spin direction, used to style the up or down button
+  // as active when up/down arrow is pressed
+  const [keyboardSpinDirection, setKeyboardSpinDirection] = React.useState(KeyboardSpinDirection.notSpinning);
+  //
```
Comment (xugao) [inline packages/react-next/src/components/SpinButton/SpinButton.base.tsx:100]: nit: remove

File: packages/react-next/src/components/SpinButton/SpinButton.base.tsx [hunk 9]
```diff
@@ -0,0 +1,452 @@
...
+    ariaValueText,
+    keytipProps,
+    className,
+    inputProps,
+    iconButtonProps,
+    styles,
+  } = props as ISpinButtonProps;
+
+  const input = React.useRef<HTMLInputElement>(null);
+  // Is true when the control has focus
```
Comment (xugao) [inline packages/react-next/src/components/SpinButton/SpinButton.base.tsx:88]: nit: remove comment, not useful

File: packages/react-next/src/components/SpinButton/SpinButton.base.tsx [hunk 10]
```diff
@@ -0,0 +1,452 @@
...
+  const [isFocused, { setTrue: setTrueIsFocused, setFalse: setFalseIsFocused }] = useBoolean(false);
+  // The value of the spin button
```
Comment (xugao) [inline packages/react-next/src/components/SpinButton/SpinButton.base.tsx:90]: nit: remove comment, not useful

File: packages/react-next/src/components/SpinButton/SpinButton.base.tsx [hunk 11]
```diff
@@ -0,0 +1,452 @@
...
+    const { precision = Math.max(calculatePrecision(calculatePrecisionProps.step!), 0) } = calculatePrecisionProps;
+    return precision;
+  };
+
+  let { value: initialValue = props.defaultValue } = props;
+  if (initialValue === undefined) {
+    initialValue = typeof min === 'number' ? String(min) : '0';
+  }
+
+  const [state] = React.useState<ISpinButtonState>({
```
Comment (xugao) [inline packages/react-next/src/components/SpinButton/SpinButton.base.tsx:113]: don't use state here. use ref

File: packages/react-next/src/components/SpinButton/SpinButton.types.ts [hunk 12]
```diff
@@ -160,14 +167,14 @@ export interface ISpinButtonProps extends React.HTMLAttributes<HTMLDivElement> {
...
-    className?: string,
-  ) => ISpinButtonClassNames;
+  // getClassNames?: (
+  //   theme: ITheme,
+  //   disabled: boolean,
+  //   isFocused: boolean,
+  //   keyboardSpinDirection: KeyboardSpinDirection,
+  //   labelPosition?: Position,
+  //   className?: string,
+  // ) => ISpinButtonClassNames;
```
Comment (xugao) [inline packages/react-next/src/components/SpinButton/SpinButton.types.ts:177]: hmmm this prop is so weird. i would love to remove this. @khmakoto ?

File: packages/react-next/src/components/SpinButton/SpinButton.types.ts [hunk 13]
```diff
@@ -263,22 +270,22 @@ export interface ISpinButtonStyles {
   /**
    * Style override when the label is positioned at the start.
    */
-  labelWrapperStart: IStyle;
+  // labelWrapperStart: IStyle;
```
Comment (xugao) [inline packages/react-next/src/components/SpinButton/SpinButton.types.ts:273]: i am good with removing these from styles prop

File: packages/react-next/src/components/SpinButton/SpinButton.test.tsx (hunk 4 above)
Comment (czearing) [reply packages/react-next/src/components/SpinButton/SpinButton.test.tsx]: I updated the code and found a problem related to this and resolved it.

File: packages/react-next/src/components/SpinButton/SpinButton.base.tsx [hunk 14]
```diff
@@ -0,0 +1,444 @@
...
+  getNativeProps,
+  divProperties,
+} from '../../Utilities';
+import { getArrowButtonStyles } from './SpinButton.styles';
+import { ISpinButtonProps, ISpinButtonStyleProps, ISpinButtonStyles, KeyboardSpinDirection } from './SpinButton.types';
+import { Position } from 'office-ui-fabric-react/lib/utilities/positioning';
+import { KeytipData } from '../../KeytipData';
+import { useBoolean, useSetTimeout, useControllableValue } from '@uifabric/react-hooks';
+
+export interface ISpinButtonState {
```
Comment (xugao) [inline packages/react-next/src/components/SpinButton/SpinButton.base.tsx:21]: remove,
```suggestion
interface ISpinButtonVariables {
```

Comment (xugao) [reply packages/react-next/src/components/SpinButton/SpinButton.base.tsx:21]: 1. "State" tends to refer to component state. but these in fact are not component state, but just local variables

File: packages/react-next/src/components/SpinButton/SpinButton.base.tsx [hunk 15]
```diff
@@ -0,0 +1,444 @@
...
+    const { precision = Math.max(calculatePrecision(calculatePrecisionProps.step!), 0) } = calculatePrecisionProps;
+    return precision;
+  };
+
+  let { value: initialValue = props.defaultValue } = props;
+  if (initialValue === undefined) {
+    initialValue = typeof min === 'number' ? String(min) : '0';
+  }
+
+  const { current: state } = React.useRef<ISpinButtonState>({
```
Comment (xugao) [inline packages/react-next/src/components/SpinButton/SpinButton.base.tsx:107]: rename `state` to `variables`

**Comment by xugao** [comment]
overall LGTM. once build passes and comments resolved, i will approve :)

File: packages/react-next/src/components/SpinButton/SpinButton.types.ts (hunk 12 above)
Comment (khmakoto) [reply packages/react-next/src/components/SpinButton/SpinButton.types.ts:177]: I think I agree with you.

**Comment by JustSlone** [comment]
Following up on older PRs to help them move along 😃

@czearing looks like there are some conflicts that need to be resolved, can you take a look?

@xugao looks like there are new screener tests to approve as part of this.

File: packages/react-next/etc/react-next.api.md [hunk 16]
```diff
@@ -1285,8 +1282,6 @@ export interface ISpinButtonProps extends React.HTMLAttributes<HTMLDivElement> {
     defaultValue?: string;
     disabled?: boolean;
     downArrowButtonStyles?: Partial<IButtonStyles>;
-    // Warning: (ae-forgotten-export) The symbol "ISpinButtonClassNames" needs to be exported by the entry point index.d.ts
-    getClassNames?: (theme: ITheme, disabled: boolean, isFocused: boolean, keyboardSpinDirection: KeyboardSpinDirection, labelPosition?: Position, className?: string) => ISpinButtonClassNames;
```
Comment (xugao) [inline packages/react-next/etc/react-next.api.md:1289]: please update notes [here](https://github.com/microsoft/fluentui/blob/master/packages/react-next/release.md) to include this breaking change

File: packages/react-next/etc/react-next.api.md [hunk 17]
```diff
@@ -1312,35 +1307,30 @@ export interface ISpinButtonProps extends React.HTMLAttributes<HTMLDivElement> {
...
+    // (undocumented)
+    labelPosition: Position;
+    // (undocumented)
+    theme: ITheme;
 }
 
 // @public (undocumented)
 export interface ISpinButtonStyles {
     arrowButtonsContainer: IStyle;
-    arrowButtonsContainerDisabled: IStyle;
```
Comment (xugao) [inline packages/react-next/etc/react-next.api.md:1324]: please update notes [here](https://github.com/microsoft/fluentui/blob/master/packages/react-next/release.md) to include this breaking change (removed props from `ISpinButtonStyles`)

File: packages/react-next/etc/react-next.api.md (hunk 16 above)
Comment (xugao) [reply packages/react-next/etc/react-next.api.md:1296]: @czearing - could you get this resolved quickly so we can merge this PR :)

**Pull request merged**